rfd/0213-relay-a-lightweight-proxy.md

```diff
@@ -0,0 +1,66 @@
+
+It should be possible for the relay reverse tunnels to support all services immediately, since no changes should be necessary for the connections tunneled from the Teleport control plane to the agents.
+
+### Client behavior
+
+The relay group address can be specified at `tsh login` time, and it can be explicitly specified and/or disabled as part of the invocation of `tsh ssh`, `tsh proxy ssh`, `tsh kube login` or `tsh proxy kube`, with either a command line option or an environment variable; as usual, the command line option overrides the envvar which overrides the configuration stored at login time. Access through machine ID will need explicit configuration as part of the "output" or "service" in use, and any need to rapidly switch between relay and non-relay routing can be served by creating two copies of a given output or service. At least in the first implementation of the client tooling side of this feature, there will be no support for automatic switching or fallback between relay and non-relay connection routing, as well as no safeguards against connecting through the broader control plane even if a relay could've been used.
+
+Specifying a relay address at login time in Teleport Connect is left for future work, but a possible way to allow that could be an "Advanced..." button in the "Enter cluster address" modal, or some UI element to change settings for the active cluster after logging in. Editing the profile configuration file is also an option (albeit one that we don't necessarily want to encourage or support), as well as using `tsh ssh --relay <relayaddr> <user>@<host>` in a terminal tab instead of choosing a server to connect to in the list of resources.
+
+All the relays of the group will serve client connections in the same way, so clients only ever need to interact directly with the single load balancer address - this is a requirement to support Kubernetes clients without a local proxy - and will do so depending on the required protocol.
+
+For SSH, the existing proxy transport protocol will also be implemented by relays, which will run the same connection router that's used by the proxy to resolve a connection target into a host ID; however, unlike the actual proxies, relays will only ever route connections to resources that are available within the same relay group and will never open TCP connections to direct dial agents, and, to avoid giving relays any more privileges than they strictly require, agentless SSH servers and proxy recording mode are not going to be supported. For consistency with the router in the Teleport proxy, host ID resolution will happen against the full SSH inventory without filtering for nodes available in the subenvironment, and the usual rules will apply in case of an ambiguous match. The list of SSH servers shown with `tsh ls` will be updated to show `<-relay (<relayname>)` rather than `<-tunnel` for nodes connected to a relay, and similar UI changes will be applied to the web UI and Connect.
```

Comment:
> The list of SSH servers shown with `tsh ls` will be updated to show `<-relay (<relayname>)` rather than `<-tunnel` for nodes connected to a relay, and similar UI changes will be applied to the web UI and Connect.

Nit: this suggests static resolution of a connection, while in practice we may want to delay that choice up until the connection attempt is made.

Reply:
The connectivity of an agent is a static property of the agent, not something decided at runtime - with the current plan of only letting the control plane speak to the agent through the relay (we might revisit this due to performance concerns), deciding whether to show an address, `<-tunnel` or `<-relay` is only a function of the node heartbeat.